src/Statement/StatementListChanger.php

```diff
@@ -0,0 +1,169 @@
+<?php
+
+namespace Wikibase\DataModel\Services\Statement;
+
+use Wikibase\DataModel\Entity\PropertyId;
+use Wikibase\DataModel\Statement\Statement;
+use Wikibase\DataModel\Statement\StatementList;
+
+/**
+ * A collection of higher level utility functions to manipulate StatementList objects.
+ *
+ * @since 3.7
+ *
+ * @license GPL-2.0+
+ * @author Thiemo Mättig
+ */
+class StatementListChanger {
+
+	/**
+	 * @var StatementList
+	 */
+	private $statementList;
+
+	public function __construct( StatementList $statementList ) {
+		$this->statementList = $statementList;
+	}
+
+	/**
+	 * Removes all statements from the $statementList object given in the constructor.
+	 */
+	public function clear() {
+		foreach ( $this->statementList->toArray() as $statement ) {
+			$this->statementList->removeStatementsWithGuid( $statement->getGuid() );
+		}
+	}
+
+	/**
+	 * Makes sure all statements with the same property are next to each other (forming a group),
+	 * and reorders them if necessary. The position of the group in the list is determined by the
+	 * first statement with the same property.
+	 */
+	public function groupByProperty() {
+		$byId = [];
+
+		foreach ( $this->statementList->toArray() as $statement ) {
+			$id = $statement->getPropertyId()->getSerialization();
+			$byId[$id][] = $statement;
+		}
+
+		$this->clear();
+
+		foreach ( $byId as $statements ) {
+			foreach ( $statements as $statement ) {
+				$this->statementList->addStatement( $statement );
+			}
+		}
+	}
+
+	/**
+	 * @param Statement $newStatement
+	 * @param int|null $index An absolute index in the list. If the index is not next to a statement
+	 *  with the same property, the closest possible position is used instead. Default is null,
```

Comment:
Yes, both directions. This only matters if the list was not grouped by property ID before.